crypto/compactcert/verifier.go

```diff
@@ -0,0 +1,96 @@
+// Copyright (C) 2019-2020 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package compactcert
+
+import (
+	"fmt"
+
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/crypto/merklearray"
+	"github.com/algorand/go-algorand/data/basics"
+)
+
+// Verifier is used to verify a compact certificate.
+type Verifier struct {
+	Params
+
+	partcom crypto.Digest
+}
+
+// MkVerifier constructs a verifier to check the compact certificate
+// on the message specified in p, with partcom specifying the Merkle
+// root of the participants that must sign the message.
+func MkVerifier(p Params, partcom crypto.Digest) *Verifier {
+	return &Verifier{
+		Params:  p,
+		partcom: partcom,
+	}
+}
+
+// Verify checks if c is a valid compact certificate for the message
+// and participants that were used to construct the Verifier.
+func (v *Verifier) Verify(c *Cert) error {
+	if c.SignedWeight < v.ProvenWeight {
+		return fmt.Errorf("cert signed weight %d < proven weight %d", c.SignedWeight, v.ProvenWeight)
+	}
+
+	// Verify all of the reveals
+	sigs := make(map[uint64]crypto.Hashable)
+	parts := make(map[uint64]crypto.Hashable)
+	for i, r := range c.Reveals {
```

Comment:
Suppose `Reveals` contains entries such that `r.Pos` is some fixed value over and over, `r.Part` is always the correct entry for that position, and `r.SigSlot` has `L`s increasing by the weight of `r.Part.Weight` each time. In other words, imagine we fill the cert with the same guy until it exceeds `v.ProvenWeight`.

Then we can build a `SigCommit` such that the `SigCommit`/`SigProofs` check succeeds (since `SigCommit` comes from the certificate), and we'll only verify the one position we chose in `v.partcom`. Then, it looks like the `hashCoin` checks should succeed too.

Where does this break?

Reply:
I think you're right: if the cert has multiple reveals for the same `r.Pos`, the code only checks the last one!  That is indeed pretty horrible.  I believe it should suffice to check that `r.Pos` is not already in `sigs`, in this first loop.  Does that sound right to you?

Reply:
I think that's right -- or maybe more intuitively (for me) that it's not already in `parts`, since `partcom` is constructed by the verifier. But I think functionality that is equivalent.